crates/storage/src/collections/slice/mod.rs

```diff
@@ -0,0 +1,384 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Slices provide (mutable) views into contiguous storage, akin to regular rust slices.
+
+mod impls;
+mod iter;
+
+use crate::{
+    collections::slice::iter::{
+        Iter,
+        IterMut,
+    },
+    lazy::{
+        LazyArray,
+        LazyIndexMap,
+    },
+    traits::PackedLayout,
+};
+use core::ops::Range;
+
+/// A view into contiguous storage.
+#[derive(Clone, Debug)]
+pub struct Slice<T> {
+    /// The start and end indices inside the `storage`. Indexing the slice using `n` means that we
+    /// access `n + range.start`.
+    range: Range<u32>,
+    /// The underlying storage structure, such as `LazyIndexMap` or `LazyArray`.
+    backing_storage: T,
+}
+
+impl<T> Slice<T>
+where
+    T: ContiguousStorage,
+{
+    /// Returns the number of elements in the slice.
+    pub fn len(&self) -> u32 {
+        self.range.end - self.range.start
+    }
+
+    /// Returns true if the slice has a length of 0.
+    pub fn is_empty(&self) -> bool {
+        self.range.end == self.range.start
+    }
+
+    /// Returns the first element of the slice, or None if it is empty.
+    pub fn first(&self) -> Option<&T::Item> {
+        self.get(0)
+    }
+
+    /// Returns the first element of the slice, or None if it is empty.
+    pub fn last(&self) -> Option<&T::Item> {
+        self.get(self.len())
+    }
+
+    /// Returns the first and all the rest of the elements of the slice, or None if it is empty.
+    pub fn split_first(&self) -> Option<(&T::Item, Slice<&T>)> {
+        let first = self.first()?;
+        Some((
+            first,
+            Slice::new(self.range.start + 1..self.range.end, &self.backing_storage),
+        ))
+    }
+
+    /// Returns the last and all the rest of the elements of the slice, or None if it is empty.
+    pub fn split_last(&self) -> Option<(&T::Item, Slice<&T>)> {
+        let first = self.last()?;
+        Some((
+            first,
+            Slice::new(self.range.start..self.range.end - 1, &self.backing_storage),
+        ))
+    }
```

Comment:
Not sure if these are really needed if we have a more general `split_at` method

Reply:
They're present in the standard library, which I've mirrored as much as possible so far.